File: srcs/hms_encoder_sec.c
/* ************************************************************************** */
/*                                                                            */
/*                                                        :::      ::::::::   */
/*   hms_encoder_sec.c                                  :+:      :+:    :+:   */
/*                                                    +:+ +:+         +:+     */
/*   By: pitriche <pitriche@student.42.fr>          +#+  +:+       +#+        */
/*                                                +#+#+#+#+#+   +#+           */
/*   Created: 2019/10/25 22:47:34 by pitriche          #+#    #+#             */
/*   Updated: 2020/01/30 11:41:00 by pitriche         ###   ########.fr       */
/*                                                                            */
/* ************************************************************************** */

#include "doom_nukem.h"

int		write_wall(t_walls *wall, int fd)
{
	unsigned char buf[32];

	ft_bzero(buf, 32);
	printf("wall > %.2f %.2f %.2f %.2f\n", wall->x2, wall->y2, wall->x1, wall->y1);
	*(signed int *)(buf + 0) = wall->x2 * 100;
	*(signed int *)(buf + 4) = wall->y2 * 100;
	*(signed int *)(buf + 8) = wall->x1 * 100;
	*(signed int *)(buf + 12) = wall->y1 * 100;
	*(unsigned short *)(buf + 16) = wall->wall_tex;
	*(unsigned short *)(buf + 18) = wall->bot_tex;
	*(unsigned short *)(buf + 20) = wall->top_tex;
	*(unsigned short *)(buf + 22) = wall->is_cross;
	*(unsigned int *)(buf + 24) = wall->sec_lnk;
	if (write(fd, buf, 32) != 32)
		return (1);
	return (0);
}

int		write_walls(t_sector *sec, int fd)
{
	unsigned int	i;

	i = 0;
	while (i < sec->nb_wal)
		write_wall(sec->walls + (i++), fd);
	return (0);
}

/*
** the + 0.001 prevents rounding errors like 2.3 * 100 being rounded to 229
*/

int		write_sector(t_sector *sec, int fd)
{
	unsigned char buf[16];
	printf("\nSECC\n");

	*(unsigned int *)(buf + 0) = (sec->fl_hei + 0.001) * 100;
	*(unsigned int *)(buf + 4) = (sec->ce_hei + 0.001) * 100;
	*(unsigned short *)(buf + 8) = sec->fl_tex;
	*(unsigned short *)(buf + 10) = sec->ce_tex;
	*(unsigned int *)(buf + 12) = sec->nb_wal;
	if (write(fd, buf, 16) != 16)
		return (1);
	if (write_walls(sec, fd))
		return (1);
	return (0);
}

int		write_sectors(t_al *al, int fd)
{
	signed int		buf[4];
	unsigned int	i;

	buf[0] = al->nb_sec;
	buf[1] = al->play.csec ? al->play.csec : 1;
	buf[2] = al->play.posx * 100;
	buf[3] = al->play.posy * 100;
	if (write(fd, buf, 16) != 16)
		return (1);
	i = 0;
	while (i++ < al->nb_sec)
		if (write_sector(al->sec + i, fd))
			return (1);
	return (0);
}

File: web version/src/hms_encoder_sec.js
// Fonction pour écrire un mur dans un buffer
function writeWall(wall) {
	const buf = new ArrayBuffer(32);
	const view = new DataView(buf);
  
	view.setInt32(0, wall.x2 * 100, true);
	view.setInt32(4, wall.y2 * 100, true);
	view.setInt32(8, wall.x1 * 100, true);
	view.setInt32(12, wall.y1 * 100, true);
	view.setUint16(16, wall.wall_tex, true);
	view.setUint16(18, wall.bot_tex, true);
	view.setUint16(20, wall.top_tex, true);
	view.setUint16(22, wall.is_cross, true);
	view.setUint32(24, wall.sec_lnk, true);
  
	return buf;
  }
  
  // Fonction pour écrire tous les murs d'un secteur
  function writeWalls(sec) {
	const buffers = [];
	for (let i = 0; i < sec.nb_wal; i++) {
	  buffers.push(writeWall(sec.walls[i]));
	}
	return buffers;
  }
  
  // Fonction pour écrire un secteur dans un buffer
  function writeSector(sec) {
	const buf = new ArrayBuffer(16);
	const view = new DataView(buf);
  
	view.setUint32(0, Math.round((sec.fl_hei + 0.001) * 100), true);
	view.setUint32(4, Math.round((sec.ce_hei + 0.001) * 100), true);
	view.setUint16(8, sec.fl_tex, true);
	view.setUint16(10, sec.ce_tex, true);
	view.setUint32(12, sec.nb_wal, true);
  
	const wallsBuffers = writeWalls(sec);
	return [buf, ...wallsBuffers];
  }
  
  // Fonction pour écrire tous les secteurs
  function writeSectors(al) {
	const headerBuf = new ArrayBuffer(32);
	const headerView = new DataView(headerBuf);
  
	headerView.setInt32(0, al.nb_sec, true);
	headerView.setInt32(4, al.play.csec || 1, true);
	headerView.setInt32(8, al.play.posx * 100, true);
	headerView.setInt32(12, al.play.posy * 100, true);
	headerView.setInt32(16, al.door, true);
	headerView.setInt32(20, al.hard, true);
	headerView.setInt32(24, (al.end_sect.x - WIN_SIZEX / 2) / 10, true);
	headerView.setInt32(28, (al.end_sect.y - WIN_SIZEY / 2) / -10, true);
  
	const buffers = [headerBuf];
	for (let i = 1; i <= al.nb_sec; i++) {
	  buffers.push(...writeSector(al.sec[i]));
	}
  
	return buffers;
  }
  
  // Fonction pour écrire les données dans un fichier
  function writeToFile(buffers, filename) {
	const blob = new Blob(buffers, { type: "application/octet-stream" });
	const link = document.createElement("a");
	link.href = URL.createObjectURL(blob);
	link.download = filename;
	link.click();
  }